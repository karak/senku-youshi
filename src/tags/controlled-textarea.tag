<controlled-textarea>
  <textarea ref="textarea" onInput={input} value={ value } rows="8"/>
  <style scoped>
  controlled-textarea textarea {
    width: 100%;
  }
  </style>
  <script>

  /* define "value" accessor connecting with textarea */
  Object.defineProperty(this, 'value', {
    enumerable: true,
    configurable: true,
    set: function (value) {
      this.refs.textarea.value = value;
    },
    get: function () {
      return this.refs.textarea.value;
    }
  });

  /** "mount" event handler */
  this.on('mount', () => {
    const { defaultValue } = this.opts;
    this.value = defaultValue || '';
  });

  /** "oninput" event handler of textarea */
  input(e) {
    const value = e.target.value;
    this.trigger('change', { value });
  }
  </script>
</controlled-textarea>
