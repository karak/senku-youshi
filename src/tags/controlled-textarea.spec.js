import { shallow } from 'riot-test-utils';
import './controlled-textarea.tag';
import { simulate } from 'simulate-event';
import { wrap } from 'module';

describe('<controlled-textarea />', () => {
  describeWithTag('refs', getWrapper => {
    it('has textarea in refs', () => {
      const wrapper = getWrapper();
      expect(wrapper.instance().refs).toHaveProperty('textarea');
      expect(wrapper.instance().refs.textarea).toBeInstanceOf(
        HTMLTextAreaElement
      );
    });
  });

  describe('opts.defaultValue', () => {
    let wrapper = null;

    afterEach(() => {
      if (wrapper) {
        wrapper.unmount();
      }
    });

    it('value is empty by default', () => {
      wrapper = shallow('controlled-textarea');

      expect(wrapper.instance().value).toBe('');
    });

    it('value is assigned from the option', () => {
      const opts = { defaultValue: 'Once upon a time...' };
      wrapper = shallow('controlled-textarea', opts);

      expect(wrapper.instance().value).toBe(opts.defaultValue);
    });
  });

  describeWithTag('value accessor', getWrapper => {
    it('assigns the set value to value of textarea', () => {
      const wrapper = getWrapper();
      wrapper.instance().value = 'Changed';
      expect(wrapper.instance().value).toBe('Changed');
      expect(wrapper.instance().refs.textarea.value).toBe('Changed');
    });

    it('get value of textarea', () => {
      const wrapper = getWrapper();
      wrapper.instance().refs.textarea.value = 'Changed';
      expect(wrapper.instance().value).toBe('Changed');
    });
  });

  describeWithTag('event "change"', getWrapper => {
    let fn;
    beforeEach(() => {
      fn = jest.fn();
    });

    afterEach(() => {
      fn = null;
    });

    it('is triggered by "input" of textarea', () => {
      const wrapper = getWrapper();

      wrapper.instance().on('change', fn);

      const { textarea } = wrapper.instance().refs;
      textarea.value = 'Changed';
      simulate(textarea, 'input');

      expect(fn).toHaveBeenCalledWith({ value: 'Changed' });
    });
  });

  it('matches snapshot', () => {
    const wrapper = shallow('controlled-textarea');
    try {
      expect(wrapper.toJSON()).toMatchSnapshot();
    } finally {
      wrapper.unmount();
    }
  });
});

function describeWithTag(suite, opts, callback) {
  if (arguments.length < 3) {
    callback = opts;
    opts = {};
  }

  describe(suite, () => {
    let wrapper = null;

    beforeEach(() => {
      wrapper = shallow('controlled-textarea', opts);
    });

    afterEach(() => {
      wrapper.unmount();
    });

    callback(() => wrapper, ...arguments);
  });
}
