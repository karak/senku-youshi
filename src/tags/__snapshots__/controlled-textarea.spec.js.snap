// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<controlled-textarea /> matches snapshot 1`] = `
<controlled-textarea
  data-is={
    Array [
      "controlled-textarea",
    ]
  }
>
  <textarea
    class={
      Array [
        "textarea",
      ]
    }
    placeholder={
      Array [
        "句を改行区切りで貼り付けてください。",
      ]
    }
    ref={
      Array [
        "textarea",
      ]
    }
    rows={
      Array [
        "20",
      ]
    }
  />
</controlled-textarea>
`;
